tests/rptest/tests/datalake/iceberg_toggling_test.py

```diff
@@ -176,3 +227,86 @@ def wait_iceberg_quiesced(*,
+                              include_query_engines=[QueryEngineType.SPARK
+                                                     ]) as dl:
+            # Create the topic with iceberg disabled
+            dl.create_iceberg_enabled_topic(self.topic_name,
+                                            partitions=NUM_PARTITIONS,
+                                            iceberg_mode="disabled")
+
+            # Misconfigure iceberg mode to value_schema_latest
+            self.set_iceberg_mode("value_schema_latest:subject=foo")
+            self.produce_avro_records(record_count=RECORD_COUNT)
+            produced_records += RECORD_COUNT
+
+            try:
+                dl.wait_for_translation(self.topic_name,
+                                        msg_count=produced_records,
+                                        timeout=30)
+                assert False, "Expected translation to fail but it did not"
+            except ducktape.errors.TimeoutError:
+                pass
+
+            self.set_iceberg_mode("value_schema_id_prefix")
+
+            dl.wait_for_translation(self.topic_name,
+                                    msg_count=produced_records,
+                                    timeout=30)
+
+    @cluster(num_nodes=3)
+    @matrix(cloud_storage_type=supported_storage_types())
+    def test_concurrent_mode_toggling(self, cloud_storage_type):
+        """
+        Test toggling iceberg mode while producing data and translation is in progress.
+        """
+        NUM_PARTITIONS = 1
+        RECORDS_PER_PRODUCE = 100
+        TOGGLES = 200
+
+        with DatalakeServices(self.test_context,
+                              self.redpanda,
+                              include_query_engines=[QueryEngineType.SPARK
+                                                     ]) as dl:
+            # Create the topic with iceberg disabled
+            dl.create_iceberg_enabled_topic(self.topic_name,
+                                            partitions=NUM_PARTITIONS,
+                                            iceberg_mode="disabled")
+            stopped = False
+            total_produced = 0
+
+            def produce_until_stopped():
+                nonlocal total_produced
+                while not stopped:
+                    self.produce_avro_records(record_count=RECORDS_PER_PRODUCE)
+                    total_produced += RECORDS_PER_PRODUCE
+                    time.sleep(0.1)
+
+            produce_t = Thread(target=produce_until_stopped, daemon=True)
+            produce_t.start()
+
+            for i in range(TOGGLES):
+                self.set_iceberg_mode("value_schema_latest:subject=foo")
+                time.sleep(0.1)
+                self.set_iceberg_mode("value_schema_id_prefix")
+
+            stopped = True
+            produce_t.join()
+
+            self.set_iceberg_mode("value_schema_id_prefix")
+            dl.wait_for_translation(self.topic_name,
+                                    msg_count=total_produced,
```

Comment:
I'm kind of surprised about this -- in toggling the mode but not toggling whether the workload is using the SR wire format, don't we expect some number of produced records to land in the DLQ?

Reply:
I was surprised too initially but I think the code intentionally classifies this as a retryable registry error, IIUC because the subject doesn't yet exist but that may be transient issue (which prompted me to name this error code https://github.com/redpanda-data/redpanda/pull/26144/commits/9b9dc2c8cf2100cd9aca01547212c978dcdbafa0#diff-c7cf15f6aa6dc8accd4d8ca3dee4387475d6ccd5c02eb384233ca58f36ff2636R26)

Now one can argue what if there is a genuine typo in the subject name in the mode and it is infact not transient which makes the case for a dlq. Perhaps we should have a more sophisticated policy that can disambiguate these cases.

Reply:
Got it, makes sense. I agree the exact feels somewhat debatable, maybe a good question for the product team. Assuming we go with this direction, we should make sure that there's a clear distinction between what's retryable and what's not. For instance https://github.com/redpanda-data/redpanda/blob/0f9a335c585dfd76cf82977de89a876b234f0102/src/v/datalake/record_schema_resolver.cc#L356 seems like maybe it should be bad_input so it's non-retryable

Reply:
ya exactly, what goes into a dlq should be discussed a bit (with input from Matt).